ledger/src/blockstore.rs

```diff
@@ -1532,22 +1564,86 @@ impl Blockstore {
         shred_index < slot_meta.consumed || data_index.contains(shred_index)
     }
 
-    fn get_data_shred_from_just_inserted_or_db<'a>(
+    /// Finds the corresponding shred at `shred_id` in the just inserted
+    /// shreds or the backing store. Panics if there is no shred.
+    fn get_shred_from_just_inserted_or_db<'a>(
         &'a self,
         just_inserted_shreds: &'a HashMap<ShredId, Shred>,
-        slot: Slot,
-        index: u64,
+        shred_id: ShredId,
     ) -> Cow<'a, Vec<u8>> {
-        let key = ShredId::new(slot, u32::try_from(index).unwrap(), ShredType::Data);
-        if let Some(shred) = just_inserted_shreds.get(&key) {
-            Cow::Borrowed(shred.payload())
-        } else {
+        let (slot, index, shred_type) = shred_id.unpack();
+        match (just_inserted_shreds.get(&shred_id), shred_type) {
+            (Some(shred), _) => Cow::Borrowed(shred.payload()),
             // If it doesn't exist in the just inserted set, it must exist in
             // the backing store
-            Cow::Owned(self.get_data_shred(slot, index).unwrap().unwrap())
+            (_, ShredType::Data) => Cow::Owned(
+                self.get_data_shred(slot, u64::from(index))
+                    .unwrap()
+                    .unwrap_or_else(|| {
+                        panic!("{} {} {:?} must be present!", slot, index, shred_type)
+                    }),
+            ),
+            (_, ShredType::Code) => Cow::Owned(
+                self.get_coding_shred(slot, u64::from(index))
+                    .unwrap()
+                    .unwrap_or_else(|| {
+                        panic!("{} {} {:?} must be present!", slot, index, shred_type)
+                    }),
+            ),
         }
     }
 
+    /// Returns true if there is no merkle root conflict between
+    /// the existing `merkle_root_meta` and `shred`
+    ///
+    /// Otherwise return false and if not already present, add duplicate proof to
+    /// `duplicate_shreds`.
+    fn perform_merkle_check(
```

Comment:
need a better name to be more clear what `true` or `false` means _at the call site_ without having to look up this comment. otherwise the usage will be very error prone.